Gaia astroquery 1.0

mhsarmiento opened the pull request:
this new pull request contain a new enhancement for the packages 'utils.tap' and 'Gaia.'  This new enhancements includes the capability to rename a user table and/or its columns.

keflavich commented on CHANGES.rst:
```diff
@@ -8,6 +8,15 @@ New Tools and Services
 Service fixes and enhancements
 ------------------------------
 
+utils.TAP
+^^^^^^^^^^^^^
```
style issue: should be
```suggestion
utils.TAP
^^^^^^^^^
```

keflavich commented on CHANGES.rst:
```diff
@@ -8,6 +8,15 @@ New Tools and Services
+
+- Added new functionality "rename_table" to rename a table and/or its columns names
+
+gaia
+^^^^^^^^^^^^^
```
```suggestion
gaia
^^^^
```

keflavich commented on astroquery/gaia/core.py:
```diff
@@ -898,5 +899,25 @@ def launch_job_async(self, query, name=None, output_file=None,
+    def rename_table(self, table_name=None, new_table_name=None, new_column_names_dict={},
+                     verbose=False):
+        """
+            This new method allows to update the column names of a user table.
+            header example: rename_table(table_name=old_table_name, new_table_name=new_table_name -optional-
+             , new_column_names_dict=[old_column1:new_column1, old_column2:new_colum2...])
+            Parameters
+            ----------
+            table_name: str, required
+                old name of the user's table
+            new_table_name: str, required
+                new name of the user's table
+            new_column_names_dict: dict str:str, required
+                dict with pairs "old_column1_name:new_column1_name"
+            verbose : bool, optional, default 'False'
+                flag to display information about the process
```
the docstring is overindented.

keflavich commented on astroquery/gaia/core.py:
```diff
@@ -898,5 +899,25 @@ def launch_job_async(self, query, name=None, output_file=None,
+                     verbose=False):
+        """
+            This new method allows to update the column names of a user table.
+            header example: rename_table(table_name=old_table_name, new_table_name=new_table_name -optional-
+             , new_column_names_dict=[old_column1:new_column1, old_column2:new_colum2...])
```
If this is meant to be an example for users to follow, it belongs in the
```
Example
-------
```
section below `Parameters`.

keflavich commented on astroquery/gaia/core.py:
```diff
@@ -898,5 +899,25 @@ def launch_job_async(self, query, name=None, output_file=None,
+    def rename_table(self, table_name=None, new_table_name=None, new_column_names_dict={},
+                     verbose=False):
+        """
+            This new method allows to update the column names of a user table.
```
Minor English correction
```suggestion
            This new method allows you to update the column names of a user table.
```

keflavich commented on astroquery/utils/tap/core.py:
```diff
@@ -1649,6 +1651,145 @@ def delete_user_table(self, table_name=None, force_removal=False,
+            else:
+                # case 2: We need to rename both, columns and column name
+                args = self.get_args_4_rename_table_all(table_name, new_table_name, new_column_names_dict)
+            # __end_if
+        # __end_if
```
remove these unnecessary comments

keflavich reviewed (changes requested):
I have a few stylistic change requests, but otherwise I think this is in good shape.

bsipocz commented:
@mhsarmiento - you could please also rebase this and address the review?

pep8speaks commented:
Hello @mhsarmiento! Thanks for updating this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:







* In the file [`astroquery/utils/tap/core.py`](https://github.com/astropy/astroquery/blob/3c70940244c1b5567c09a0aaed00a265ba764d05/astroquery/utils/tap/core.py):

> [Line 2167:38](https://github.com/astropy/astroquery/blob/3c70940244c1b5567c09a0aaed00a265ba764d05/astroquery/utils/tap/core.py#L2167): [W292](https://duckduckgo.com/?q=pep8%20W292) no newline at end of file

bsipocz commented on docs/gaia/gaia.rst:
```diff
@@ -940,6 +933,26 @@ will be able to access to your shared table in a query.
+    >>> from astroquery.utils.tap.core import TapPlus, TAP_CLIENT_ID
+    >>> from astroquery.utils.tap import taputils
+    >>> gaia = GaiaClass(gaia_tap_server='https://<env>.esac.esa.int/', gaia_data_server='https://<env>.esac.esa.int/')
+    >> >gaia.login()
```
minor typo
```suggestion
    >>> gaia.login()
```

mhsarmiento replied:
Suggested change applied

bsipocz commented on docs/gaia/gaia.rst:
```diff
@@ -940,6 +933,26 @@ will be able to access to your shared table in a query.
+    >>> from astroquery.utils.tap.model.tapcolumn import TapColumn
+    >>> from astroquery.utils.tap.core import TapPlus, TAP_CLIENT_ID
+    >>> from astroquery.utils.tap import taputils
+    >>> gaia = GaiaClass(gaia_tap_server='https://<env>.esac.esa.int/', gaia_data_server='https://<env>.esac.esa.int/')
```
what is `<env>`? I don't see mentioned anywhere above (unlike `user_<user_login_name>`). I think an example would be useful.

mhsarmiento replied:
Example updated.

Again thank you very much for your time and suggestions.

bsipocz reviewed (commented):
It appears that I had a review started a while ago but it hasn't been submitted yet, I'm sorry for the delay.

There are a couple of comments, most of them are minor.

eerovaher commented on astroquery/utils/tap/core.py:
```diff
@@ -1648,6 +1652,131 @@ def delete_user_table(self, table_name=None, force_removal=False,
+                                                 verbose,
+                                                 200)
+        msg = f"Table '{table_name}' updated."
+        print(msg)
```
Shouldn't this only be printed if the user has set `verbose` to be `True`?

mhsarmiento replied:
Yes, that's correct! Suggested change applied. Thank you!

eerovaher commented on astroquery/utils/tap/core.py:
```diff
@@ -1648,6 +1652,131 @@ def delete_user_table(self, table_name=None, force_removal=False,
+        new_column_names = ""
+        # check if the changes proposed for the columns are correct.
+        for key, value in new_column_names_dict.items():
+            if key == '':
+                raise ValueError(f" Old column name introduced "f"{key}"f" cannot be empty")
```
Instead of checking if `key == ''` wouldn't it be better to check if `key` is actually present in the table? Or is that not possible?

eerovaher commented on astroquery/utils/tap/core.py:
```diff
@@ -1648,6 +1652,131 @@ def delete_user_table(self, table_name=None, force_removal=False,
+        for key, value in new_column_names_dict.items():
+            if key == '':
+                raise ValueError(f" Old column name introduced "f"{key}"f" cannot be empty")
+            if key == value:
+                raise ValueError(f" Old column name introduced "f"{key}"f" and new column name introduced "
+                                 f" "f"{key}"f" cannot be the same")
```
Wouldn't it be better to check if the new column name is already in the table or not? Or is that not possible?

eerovaher commented on astroquery/utils/tap/core.py:
```diff
@@ -1648,6 +1652,131 @@ def delete_user_table(self, table_name=None, force_removal=False,
+                raise ValueError(f" Old column name introduced "f"{key}"f" and new column name introduced "
+                                 f" "f"{key}"f" cannot be the same")
+            if value == "":
+                raise ValueError(f" New column name introduced "f"{value}"f" cannot be empty")
```
```suggestion
                raise ValueError(f"New name of {key} cannot be an empty string")

```

eerovaher commented on astroquery/utils/tap/core.py:
```diff
@@ -1648,6 +1652,131 @@ def delete_user_table(self, table_name=None, force_removal=False,
+                raise ValueError(f" New column name introduced "f"{value}"f" cannot be empty")
+
+            # Converting dict into a string with the format expected by the TAP server
+            new_pair = key + ":" + value
+            new_column_names = new_column_names + new_pair
+            count = count + 1
+            if count < len(new_column_names_dict):
+                new_column_names = new_column_names + ','
```
Assuming that the preceding code has verified all the data then the following would be more pythonic:
```suggestion
        new_column_names = ','.join(f'{key}:{value}' for key, value in new_column_names_dict.items())
```

mhsarmiento replied:
Thank you very much. Thank to you guys I'm learning and improving my knowledge of python. Thanks!

eerovaher commented on astroquery/utils/tap/core.py:
```diff
@@ -1648,6 +1652,131 @@ def delete_user_table(self, table_name=None, force_removal=False,
+        }
+        return args
+
+    def get_args_4_rename_table_only_columns(self, table_name, new_column_names_dict):
```
The only difference between this function and `get_args_4_rename_table_all()` is that this one does not include `"new_table_name"` in `args`. I don't see why it would be desirable to implement two functions that duplicate each other to such an extent.

mhsarmiento replied:
Suggested change applied

eerovaher reviewed (commented):
I have a few comments and suggestions, but also a few questions.

mhsarmiento commented on astroquery/gaia/core.py:
```diff
@@ -11,7 +11,8 @@
 Created on 30 jun. 2016
-
+Modified on 1 jun. 2021 by mhsarmiento
+Version: gaia-astroquery-1.0
```
Change applied

mhsarmiento commented on astroquery/gaia/core.py:
```diff
@@ -898,5 +899,28 @@ def launch_job_async(self, query, name=None, output_file=None,
 
+    def rename_table(self, table_name=None, new_table_name=None, new_column_names_dict={},
+                     verbose=False):
+        """ This new method allows you to update the column names of a user table.
```
Change applied. Thanks!

mhsarmiento commented on astroquery/gaia/core.py:
```diff
@@ -898,5 +899,28 @@ def launch_job_async(self, query, name=None, output_file=None,
+
+        Parameters
+        ----------
+        table_name: str, required
```
Change applied. Thanks!

mhsarmiento commented on astroquery/utils/tap/core.py:
```diff
@@ -1649,6 +1649,132 @@ def delete_user_table(self, table_name=None, force_removal=False,
+        args = {}
+
+        if table_name is None:
+            raise ValueError("Table name cannot be null")
```
Change applied. Thanks!

mhsarmiento commented on astroquery/utils/tap/core.py:
```diff
@@ -1649,6 +1649,132 @@ def delete_user_table(self, table_name=None, force_removal=False,
+        if table_name is None:
+            raise ValueError("Table name cannot be null")
+        if (new_table_name is None or new_table_name == '') and \
+                (new_column_names_dict is None or not new_column_names_dict):
```
Suggested change applied. Thanks!

mhsarmiento commented on astroquery/gaia/core.py:
```diff
@@ -889,9 +925,31 @@ def launch_job_async(self, query, name=None, output_file=None,
+            if output_format in format_with_results_compressed:
+                # In this case we will have to take also into account the .fits format
+                if not output_file.endswith(compressed_extension):
+                    warnings.warn('WARNING!!! By default, results in "votable" and "fits" format are returned in '
```
Suggested change applied. Thanks!

mhsarmiento commented on astroquery/utils/tap/tests/test_tap.py:
```diff
@@ -955,6 +953,50 @@ def test_update_user_table():
+                           context=None,
+                           body=tableData,
+                           headers=None)
+    # tableRequest = f"tables?tables={tableName}"
```
Suggested change applied. Thanks!

mhsarmiento commented on astroquery/utils/tap/tests/test_tap.py:
```diff
@@ -955,6 +953,50 @@ def test_update_user_table():
+    # data = connHandler.url_encode(dictArgs)
+    # response = connHandler.execute_table_tool(data, verbose=verbose)
+
+    with pytest.raises(Exception):
```
Not sure yet which type of exceptions are risen here. I have already open an issue on my side to improve this part of the code in future versions. Thank you very much for pointing this out.

bsipocz replied:
Having an issue open in the astroquery repo may be useful, too, so we see what has come up already.

mhsarmiento commented on docs/gaia/gaia.rst:
```diff
@@ -77,13 +77,6 @@ Examples
-
-  >>> Gaia.MAIN_GAIA_TABLE = "gaiadr2.gaia_source"  # Select Data Release 2, default
-  >>> Gaia.MAIN_GAIA_TABLE = "gaiaedr3.gaia_source" # Select early Data Release 3
-
```
Sorry for this. Was due to a problem during the rebase. I have restored the example.

mhsarmiento commented on astroquery/gaia/core.py:
```diff
@@ -842,8 +854,30 @@ def launch_job(self, query, name=None, output_file=None,
+                if not output_file.endswith(compressed_extension):
+                    warnings.warn('WARNING!!! By default, results in "votable" and "fits" format are returned in '
+                                  f'compressed format therefore your file {output_file} '
+                                  f'will be renamed to {output_file}.gz')
```
It is from the TAP+ protocol. By default certain type of output formats are returned in compress format

eerovaher replied:
If it's a feature of the TAP+ protocol then it would be better to open a separate pull request for implementing this in the `TapPlus` class.

mhsarmiento replied:
Sorry, I'm not sure if I understand here what do you want me to do. TAP+ returning a compressed format is not something that has been implemented now. 
The problem here was that when a user gives a name, for example "myVotable.vot" for the file and he/she does not expect that the file is going to be provided in compressed format and then he/she tries to open the file, the user is going to get and error because the file is compressed. In order to avoid that, we were requested to rename the file with the correct extension if the user didn't provided that at first 'myVotable.vot.gz', so they don't get that error when opening the file.

eerovaher replied:
If the `votable` and `fits` files are compressed only by the Gaia archive then this code belongs here. But from your previous answer I got the impression that those file types being compressed is required by the TAP+ standard, and if that is the case then this code needs to go to the `TapPlus` class so that all `astroquery` modules connecting to TAP+ services could benefit from it.

mhsarmiento replied:
It required by TAP+. I can move it to TapPlus for the next implementation but we have dealing with this pull request since past June and I/we would really appreciate if we can move on with this. Specially if as you have suggested, we need to do a different pull request for the TAP+ modifications.  I will take this also into account from now on.

eerovaher replied:
This pull request would implement many improvements to `astroquery.gaia`, but at least a few of them seem to be independent from the others. The problem with such a large pull request is that it can only be merged all at once, so if any of the proposed improvements requires some more work then merging all the other improvements has to be delayed too. Opening a separate pull request for each independent improvement would preferable.

jespinosaar replied:
Dear @eerovaher , @mhsarmiento ,

I think there is a misunderstanding here. Considering previous comments, what you are proposing is that we include that warning as a check in the launch_job method from TAPPlus class, so the rest of our modules could benefit from it. Is that correct? And that this improvement should be included in a different pull request, am I wrong?

If that is the case, I propose this: 

- On Monday, I will extract this piece of code.
- I will create a new branch from astroquery main to include only this feature.
- After testing it and checking that it works for all ESA modules, I will create the pull request. As this is something simple (just a few lines), it should be merged fast.
- Once it is accepted and in astroquery main branch, we will rebase this PR branch to include it and remove the code from Gaia module.

What do you think?

Have a nice weekend!

eerovaher replied:
My understanding based on the answers above is that changing file name extensions of compressed formats would be useful for all TAP+ services, and in that case it should be implemented in the `TapPlus` class to avoid code repetition. But I am not familiar with the TAP+ standard so it is possible I have misunderstood something. If I am mistaken and this code is useful only for the Gaia archive then it should stay where it is.

If the code in question does belong in `TapPlus` then I would say it would have been better to open a separate pull request. But the code is already included here, so opening another pull request at this point might not be worth the effort. It is best to ask @keflavich, @bsipocz or @ceb8 what they think.

> As this is something simple (just a few lines), it should be merged fast.

I do not have write access to this repository, so I cannot make any promises about when this or any other pull request gets merged, but I would expect that smaller and simpler pull requests get merged faster.

mhsarmiento commented on astroquery/gaia/core.py:
```diff
@@ -900,5 +958,28 @@ def launch_job_async(self, query, name=None, output_file=None,
                                         upload_table_name=upload_table_name,
                                         autorun=autorun)
 
+    def rename_table(self, table_name=None, new_table_name=None, new_column_names_dict={},
```
Suggested change applied. Thanks!

mhsarmiento commented on astroquery/utils/tap/conn/tapconn.py:
```diff
@@ -87,7 +87,7 @@ def __init__(self, ishttps,
         self.__connPortSsl = sslport
         if server_context is not None:
-            if(server_context.startswith("/")):
+            if (server_context.startswith("/")):
```
Suggested change applied. Thanks!

mhsarmiento commented on astroquery/utils/tap/conn/tapconn.py:
```diff
@@ -107,7 +107,7 @@ def __init__(self, ishttps,
     def __create_context(self, context):
         if (context is not None and context != ""):
-            if(str(context).startswith("/")):
+            if (str(context).startswith("/")):
```
Suggested change applied. Thanks!

mhsarmiento commented on astroquery/utils/tap/core.py:
```diff
@@ -1648,6 +1652,131 @@ def delete_user_table(self, table_name=None, force_removal=False,
+                (new_column_names_dict is None or not new_column_names_dict):
+            raise ValueError("Please introduce as minimum a new table tame or a new name for a column with format "
+                             "old_column1_name:new_column1_name, ... ,old_columnN_name:new_columnN_name")
+        if new_table_name is not None or new_table_name != '':
```
logic reviewed. Thank you for this suggestion.

mhsarmiento commented on astroquery/utils/tap/core.py:
```diff
@@ -1648,6 +1652,131 @@ def delete_user_table(self, table_name=None, force_removal=False,
+            raise ValueError("Please introduce as minimum a new table tame or a new name for a column with format "
+                             "old_column1_name:new_column1_name, ... ,old_columnN_name:new_columnN_name")
+        if new_table_name is not None or new_table_name != '':
+            if new_column_names_dict is None or not new_column_names_dict:
```
Suggested change applied. Thanks!

bsipocz commented on astroquery/utils/tap/xmlparser/utils.py:
```diff
@@ -24,11 +24,18 @@ def util_create_string_from_buffer(buffer):
+
+    try:
+        result = APTable.read(io.BytesIO(gzip.decompress(data.read())), format=astropy_format)
+    except gzip.BadGzipFile:
```
We still support python3.7, and `BadGzipFile` has only been added into 3.8, so a workaround is needed here. I see two ways, but maybe there are others (I don't have a preference of which one you do):

1) - except on `OSError` as `BadGzipFile` inherits for that one. Based on the docs, `EOFError` or `zlib.error` could be options, too.

2) Add a custom exception based on the python version. Somewhat ugly, but it is the closest to what you're checking right now, and would allow a trivial cleanup once 3.7 support is dropped.
```
try:
    # We need python > 3.7 for this
    from gzip import BadGzipFile
except ImportError:
    class BadGzipFile(OSError):
        pass
```

mhsarmiento replied:
Thank you very much. Working on this right now

mhsarmiento replied:
change applied. thank you again for the suggestion. very useful

bsipocz commented on astroquery/utils/tap/xmlparser/utils.py:
```diff
@@ -44,14 +51,18 @@ def read_http_response(response, outputFormat, correct_units=True):
+        return 'ascii.ecsv'
+    elif 'csv' == output_format:
+        return 'ascii.csv'
+    elif 'votable_plain' == output_format:
+        return 'ascii'
```
Would this be a simple ascii table, or votable format? If the latter why not use `votable`?

eerovaher replied:
Can you provide a minimal code example?

mhsarmiento replied:
Sorry, but with type of example? Do you want the JupiterNotebook with the error returned by astropy table if we don't do that conversion? Honestly I would not like to rollback the code just to show this but I have to, I will do it of course.

mhsarmientoc replied:
Ok. change applied. Thanks!

bsipocz commented:
@mhsarmiento - I've rebased and removed the duplicated commits, and now it became quite visible that the first few commits were made by someone who didn't set their email address to be recognized by github. Please let me know if you would like to fix commits by `ciwuser <ciwuser@mail.com>`, and if yes what name and email address should I amend those commits. If we do that before merging it will get into the git history properly, after merge it's a bit more complicated to fix (and won't be really fixed on the commit level). Thanks!

bsipocz reviewed (approved):
This can be merged once my minor last question about contributor git history credits are answered.

mhsarmiento commented:
> @mhsarmiento - I've rebased and removed the duplicated commits, and now it became quite visible that the first few commits were made by someone who didn't set their email address to be recognized by github. Please let me know if you would like to fix commits by `ciwuser <ciwuser@mail.com>`, and if yes what name and email address should I amend those commits. If we do that before merging it will get into the git history properly, after merge it's a bit more complicated to fix (and won't be really fixed on the commit level). Thanks!

Dear @bsipocz ,I have already fix the commits by `ciwuser` Thank you very for your work on this!

bsipocz commented:
Thank you @mhsarmiento!
